### PR TITLE
Added XNNPACK v0.0.20200225

### PR DESCRIPTION
**Pull request by stemann**

### CODE REVIEW

**Comment by stemann** [comment]
Windows excluded due to lack of support - even with a HACK to enable the build: https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=17939&view=logs&j=69ba10d4-6eef-5253-f903-1e4fed154e5e&t=fbf4254e-36bf-5ebb-6114-2c41f6e01571&l=385

**Comment by stemann** [comment]
PowerPC64LE excluded due to lack of support - even with a hack to enable the build: https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=17940&view=logs&j=21b4c91a-50f0-5731-36e9-f97674da517e&t=717d4a3c-80cf-562e-4b25-0e37f504bb7f&l=511

File: X/XNNPACK/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms()
+filter!(p -> !Sys.isfreebsd(p), platforms) # FreeBSD is unsupported
+filter!(p -> !Sys.iswindows(p), platforms) # Windows is unsupported
+filter!(p -> arch(p) != "powerpc64le", platforms) # PowerPC64LE is unsupported
+platforms = vcat(
+    filter(p -> arch(p) != "aarch64", platforms),
+    [Platform(arch(p), os(p); march = "armv8.2-a+fp16") for p in filter(p -> arch(p) == "aarch64", platforms)]
```
Comment (giordano) [inline X/XNNPACK/build_tarballs.jl:56]: Microarchitectures stuff is broken at the moment, don't use it https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/205

File: X/XNNPACK/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using BinaryBuilderBase: os
+
+name = "XNNPACK"
+version = v"0.0.20200225"
+
+# Collection of sources required to complete build
+sources = [
+    GitSource("https://github.com/google/XNNPACK.git", "7493bfb9d412e59529bcbced6a902d44cfa8ea1c"),
+    DirectorySource("./bundled"),
+    GitSource("https://github.com/pytorch/cpuinfo.git", "d5e37adf1406cf899d7d9ec1d317c47506ccb970"),
```
Comment (giordano) [inline X/XNNPACK/build_tarballs.jl:13]: Sigh

File: X/XNNPACK/bundled/patches/xnnpack-disable-fast-math.patch [hunk 3]
```diff
@@ -0,0 +1,12 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 40cafbfbd..17374f6f5 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -1499,7 +1499,6 @@ ELSE()
+ ENDIF()
+ IF(NOT CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
+   SET_PROPERTY(SOURCE ${XNNPACK_PSIMD_ACCMATH_MICROKERNEL_SRCS} APPEND_STRING PROPERTY COMPILE_FLAGS " -O3 ")
+-  SET_PROPERTY(SOURCE ${XNNPACK_PSIMD_FASTMATH_MICROKERNEL_SRCS} APPEND_STRING PROPERTY COMPILE_FLAGS " -O3 -ffast-math ")
```
[the remark this answers is not in the document]
Comment (stemann) [reply X/XNNPACK/bundled/patches/xnnpack-disable-fast-math.patch:9]: Oh yes - misread the line before it (ACCMATH/FASTMATH) as being the same target. Fixed

File: X/XNNPACK/build_tarballs.jl (hunk 2 above)
Comment (stemann) [reply X/XNNPACK/build_tarballs.jl:12]: Yeah, I guess there's no way around it for these PyTorch dependencies...

File: X/XNNPACK/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+    Dependency("CPUInfo_jll", v"0.0.20200122"),
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+    preferred_gcc_version=v"9",
```
Comment (stemann) [inline X/XNNPACK/build_tarballs.jl:66]: Not completely sure about the required GCC version.

More recent versions of XNNPACK use v9.3: https://github.com/google/XNNPACK/runs/5426900879?check_suite_focus=true

File: X/XNNPACK/build_tarballs.jl (hunk 1 above)
Comment (stemann) [reply X/XNNPACK/build_tarballs.jl]: OK

File: X/XNNPACK/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libXNNPACK", :libxnnpack),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+    Dependency("CPUInfo_jll", v"0.0.20200122"),
+]
```
Comment (stemann) [inline X/XNNPACK/build_tarballs.jl:62]: Should add PThreadPool_jll once #4559 is merged.

File: X/XNNPACK/build_tarballs.jl (hunk 1 above)
Comment (stemann) [reply X/XNNPACK/build_tarballs.jl]: Seems like there would be a substantial benefit when it is possible to build specific versions for AVX512 etc: https://github.com/google/XNNPACK#supported-architectures

File: X/XNNPACK/build_tarballs.jl (hunk 4 above)
Comment (stemann) [reply X/XNNPACK/build_tarballs.jl]: v5 seems to work

File: X/XNNPACK/build_tarballs.jl (hunk 5 above)
Comment (stemann) [reply X/XNNPACK/build_tarballs.jl:63]: Done

File: X/XNNPACK/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libXNNPACK", :libxnnpack),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+    Dependency("CPUInfo_jll", v"0.0.20200122"),
+    Dependency("PThreadPool_jll", v"0.0.20191029"),
```
Comment (giordano) [inline X/XNNPACK/build_tarballs.jl:62]: If the version number is important, you may want to specify that as a compat:
```suggestion
    Dependency("CPUInfo_jll"; compat="0.0.20200122"),
    Dependency("PThreadPool_jll"; compat="0.0.20191029"),
```
which is recorded also in the `Project.toml` of the package

**Pull request merged**